internal/provider/aws/provider.go

```diff
@@ -579,118 +585,162 @@ func (p *awsProvider) ReconcileFirewallRules(
 			"securityGroupID", securityGroupID,
 			"networkInterfaceID", networkInterface.NetworkInterfaceId,
 		)
 	}
 
-	conditions = append(
-		conditions,
-		kmetav1.Condition{
-			Type:    v1alpha1.FirewallRuleConditionTypeNetworkInterfaceAssociated,
-			Status:  kmetav1.ConditionTrue,
-			Reason:  v1alpha1.FirewallRuleConditionReasonNetworkInterfaceAssociated,
-			Message: fmt.Sprintf("Security group associated with network interface %s", aws.StringValue(networkInterface.NetworkInterfaceId)),
-		},
-	)
+	meta.SetStatusCondition(&status.Conditions, kmetav1.Condition{
+		Type:               v1alpha1.FirewallRuleConditionTypeNetworkInterfaceAssociated,
+		Status:             kmetav1.ConditionTrue,
+		ObservedGeneration: firewallRule.Generation,
+		Reason:             v1alpha1.FirewallRuleConditionReasonNetworkInterfaceAssociated,
+		Message:            fmt.Sprintf("Security group associated with network interface %s", aws.StringValue(networkInterface.NetworkInterfaceId)),
+	})
 
 	status.InstanceID = &instanceID
 	status.NetworkInterfaceID = networkInterface.NetworkInterfaceId
 
-	frSpecs := provider.EncodeFirewallRuleSpecs(firewallRules)
+	frSpec := provider.EncodeFirewallRuleSpec(firewallRule)
+	frSpecs := provider.EncodeFirewallRuleSpecs(firewallrules)
+
+	// Revoke useless permissions.
+	// Rules in the security group, that are not in the firewall rules list are considered useless.
+	for _, e := range converter.DecodeIpPermissions(securityGroup.IpPermissions) {
+		if !provider.ContainsPermission(provider.GetIngressIPPermissions(frSpecs), e) {
+			if err := p.revokeSecurityGroupIngress(ctx, log, securityGroupID, *e); err != nil {
+				return status, fmt.Errorf("failed to delete useless ingress permission: %w", err)
+			}
+		}
+	}
+	for _, e := range converter.DecodeIpPermissions(securityGroup.IpPermissionsEgress) {
+		if !provider.ContainsPermission(provider.GetEgressIPPermissions(frSpecs), e) {
+			if err := p.revokeSecurityGroupEgress(ctx, log, securityGroupID, *e); err != nil {
+				return status, fmt.Errorf("failed to delete useless egress permission: %w", err)
+			}
+		}
+	}
+
+	if !firewallRule.DeletionTimestamp.IsZero() {
+		if frSpec.Direction == provider.DirectionIngress &&
+			provider.ContainsPermission(converter.DecodeIpPermissions(securityGroup.IpPermissions), frSpec.IPPermission) {
+			// Revoke Ingress permission reconciliation
+			if err := p.revokeSecurityGroupIngress(ctx, log, securityGroupID, *frSpec.IPPermission); err != nil {
+				return status, fmt.Errorf("failed to delete security group ingress permission: %w", err)
+			}
+			if len(securityGroup.IpPermissions) == 1 && len(securityGroup.IpPermissionsEgress) == 0 {
+				// If the security group has only one ingress rule and no egress rule, delete the security group
+				if err := p.ReconcileFirewallRulesDeletion(ctx, log, nodeName); err != nil {
+					return status, fmt.Errorf("failed to to reconcile FirewallRule deletion: %w", err)
+				}
+			}
```

Comment:
If it is the last permission, why do you revoke it before deleting the security group ?